sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/StringFormatArgumentNumberMismatch.cs

```diff
@@ -18,146 +18,326 @@
+        };
 
         protected override void Initialize(SonarAnalysisContext context)
         {
-            context.RegisterSyntaxNodeActionInNonGenerated(
-                c =>
+            context.RegisterSyntaxNodeActionInNonGenerated(CheckForInvalidFormatString,
+                SyntaxKind.InvocationExpression);
+        }
+
+        private void CheckForInvalidFormatString(SyntaxNodeAnalysisContext c)
+        {
+            var invocation = (InvocationExpressionSyntax)c.Node;
+            var methodSymbol = c.SemanticModel.GetSymbolInfo(invocation).Symbol as IMethodSymbol;
+
+            if (methodSymbol == null ||
+                !IsHandledFormatMethod(methodSymbol))
+            {
+                return;
+            }
+
+            var formatArgumentIndex = methodSymbol.Parameters[0].IsType(KnownType.System_IFormatProvider) ? 1 : 0;
+            var formatExpression = invocation.ArgumentList.Arguments[formatArgumentIndex];
+
+            var constValue = c.SemanticModel.GetConstantValue(formatExpression.Expression);
+            if (!constValue.HasValue)
+            {
+                return; // can't check non-constant format strings
+            }
+
+            string formatString = (string)constValue.Value;
+
+            try
+            {
+                StringFormat.Parse(formatString)
+                            .Validate(invocation.ArgumentList, formatArgumentIndex, c.SemanticModel);
+            }
+            catch (FormatException ex)
+            {
+                c.ReportDiagnostic(Diagnostic.Create(Rule, invocation.GetLocation(), ex.Message));
+                return;
+            }
+
+            // Safety-net to ensure we haven't missed any invalid case.
+            if (!IsFormatValid(formatString))
+            {
+                c.ReportDiagnostic(Diagnostic.Create(Rule, invocation.GetLocation(),
+                    "the format is malformed and will likely throw a 'FormatException'."));
+            }
+        }
+
+        private bool IsHandledFormatMethod(IMethodSymbol methodSymbol)
+        {
+            return HandledFormatMethods.Where(hfm => methodSymbol.ContainingType.Is(hfm.ContainingType))
+                .Any(method => method.Name == methodSymbol.Name);
+        }
+
+        internal static bool IsFormatValid(string format, int argsCount = 1000000)
+        {
+            try
+            {
+                string.Format(format, new object[argsCount]);
+                return true;
+            }
+            catch (FormatException)
+            {
+                return false;
+            }
+        }
+
+        public class StringFormat
+        {
+            public class StringFormatItem
+            {
+                internal StringFormatItem(int index, int? alignment, string formatString)
+                {
+                    Index = index;
+                    Alignment = alignment;
+                    FormatString = FormatString;

```

Comment:
![CRITICAL](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-critical.png 'Severity: CRITICAL') Remove or correct this useless self-assignment. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=csharpsquid%3AS1656)